AWS OIDC Support

cletustboone opened the pull request:
# APN compliance issue.

Related to [this PR](https://github.com/Epoxy-ai/epoxy-developer-portal/pull/132).

In all the shared workflows that interact with AWS, we must...

Remove these required params:

```yaml
    secrets:
      aws_access_key_id:
        required: true
      aws_secret_access_key:
        required: true
```

And replace with:

```yaml
    inputs:
      region:
        required: true
        type: string
      aws_role_to_assume:
        required: true
        type: string
      aws_role_duration_seconds:
        required: false
        type: number
        default: 1200
```

Then later, in the workflow, replace this:
```yaml
      - name: Configure AWS CLI
        uses: aws-actions/configure-aws-credentials@v1-node16
        with:
          aws-access-key-id: ${{ secrets.aws_access_key_id }}
          aws-secret-access-key: ${{ secrets.aws_secret_access_key }}
          aws-region: ${{ inputs.region }}
```

With this:

```yaml
    - name: Configure AWS CLI
      uses: aws-actions/configure-aws-credentials@v1-node16
      with:
        role-to-assume: ${{ inputs.aws_role_to_assume }}
        role-duration-seconds: ${{ inputs.aws_role_duration_seconds }} 
        aws-region: ${{ inputs.region }}
```

cletustboone committed:
feat: Deploy cdk uses temp aws creds

cletustboone committed:
feat: cf outputs uses temp aws credentials

cletustboone committed:
fix: Add job workflow permissions for oidc

aaron-epoxy commented on .github/workflows/deploy_with_cdk.yml:
```diff
@@ -9,14 +12,17 @@ on:
-      aws_secret_access_key:
-        required: true
     inputs:
       region:
```
I think it makes sense to remove the inconsistencies in the use of `region` vs. `aws-region` at this time, since we're going to be updating all deployments in other projects.

cletustboone replied:
So would you rather use `region` or `aws-region`? Do I need to change something here?

aaron-epoxy replied:
I must have gotten my wires crossed between the `inputs` in the definition of a reusable workflow and the `with` param inside an individual job within a workflow. This is not an issue. Carry on.

aaron-epoxy commented:
Looks good. Seems like this is the rollout process:
1. Update any projects that use [this code](https://github.com/Epoxy-ai/audience-cloud-ingest/blob/main/.github/workflows/push_deploy.yml#L22-L29) to point to this `aws-oidc` branch with new secrets fields.
2. Only merge this PR once all projects have been updated and are working.
3. Eventually update all of the deployment projects to point back to `main` of this repo...which we'll almost certainly forget to do.

Also, will the policies and whatnot that make OIDC possible be created in the backplane cdk project? If so, please link the relevant code for posterity.

🚀

cletustboone commented:
Good question re: the policies that make this possible. Here is [the AWS documentation](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-idp_oidc.html#idp_oidc_Create_GitHub) that explains what policy needs to be added.

However, I have changed it slightly to look like this:
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Federated": "arn:aws:iam::220835350133:oidc-provider/token.actions.githubusercontent.com"
            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringLike": {
                    "token.actions.githubusercontent.com:sub": [
                        "repo:Epoxy-ai/epoxy-widget:ref:refs/heads/deploy/parx-prod",
                        "repo:Epoxy-ai/audience-cloud-ingest:ref:refs/heads/deploy/parx-prod",
                        "repo:Epoxy-ai/audience-cloud-ingest:ref:refs/heads/deploy/rsi-prod",
                        "repo:Epoxy-ai/epoxy-developer-portal:ref:refs/heads/dev"
                    ],
                    "token.actions.githubusercontent.com:aud": "sts.amazonaws.com"
                }
            }
        }
    ]
}
```
Allowing actions from any branch of any epoxy-ai repo to do a deploy is too risky. That is why the policy above defines the specific branches of specific repos that are allowed to deploy. As new repos get added, this must be updated. It's certainly possible to set this policy up through CDK in the backplane repo. It will need to be actively managed though.

cletustboone committed:
fix: Adjust inputs of reusable jobs

cletustboone committed:
fix: Adjust params of job some more

cletustboone committed:
fix: Correct reference to aws region

cletustboone committed:
feat: Upload to s3 uses OIDC

cletustboone committed:
ci: Update deploy_with_cdk.yml permissions

Allow read access to the repo.

cletustboone committed:
ci: Remove permissions from deploy_with_cdk.yml

The repo -> settings -> actions -> general -> allow read and write to the gh token by default.

cletustboone committed:
ci: Revert removing permissions

Now seeing issues with authenticating. Think we need write access for the id-token scope and read access for the contents scope. Fingers crossed.

cletustboone committed:
ci: Adds permissions to upload_artifact_to_s3.yml

cletustboone committed:
ci: Update upload_file_to_s3.yml to use OIDC

aaron-epoxy commented:
support for OIDC merged into main with #14
